src/utils/memory_pool.hpp

```diff
@@ -0,0 +1,274 @@
+// Alamos National Laboratory (LANL), which is operated by Triad National Security, LLC
+// for the U.S. Department of Energy/National Nuclear Security Administration. All rights
+// in the program are reserved by Triad National Security, LLC, and the U.S. Department
+// of Energy/National Nuclear Security Administration. The Government is granted for
+// itself and others acting on its behalf a nonexclusive, paid-up, irrevocable worldwide
+// license in this material to reproduce, prepare derivative works, distribute copies to
+// the public, perform publicly and display publicly, and to permit others to do so.
+//========================================================================================
+
+#ifndef UTILS_MEMORY_POOL_HPP_
+#define UTILS_MEMORY_POOL_HPP_
+
+#include <iostream>
+#include <Kokkos_Core.hpp>
+#include <string>
+#include <vector>
+#include <iostream>
+#include <memory>
+#include <stack>
+#include <map>
+#include <unordered_map>
+#include <type_traits>
+#include <math.h>
+#include <tuple>
+#include <utility>
+
+namespace parthenon { 
+
+// Object for managing a pool of Kokkos::Views that
+// have the same instantiation call signature
+template <class T, class KEY_T = unsigned long int>
+class Pool
+{
+public:
+  using base_t = T;
+  // Forward declarations of pool types
+  class weak_t;
+  class owner_t;
+
+private:
+  std::function<T()> get_resource_;
+  std::stack<weak_t> available_;
+  std::unordered_map<KEY_T, std::pair<weak_t, int>> inuse_;
+  static const KEY_T default_key_ = KEY_T();
+  KEY_T keyc_;
+
+public:
+  template <class... Ts>
+  Pool(std::function<T()> get_resource) : get_resource_(get_resource),
+                                          available_(),
+                                          inuse_(),
+                                          keyc_(default_key_) {}
+
+  weak_t Get();
+
+  void PrintStatistics()
+  {
+    std::cout << available_.size() << " unused objects." << std::endl;
+    std::cout << inuse_.size() << " used objects." << std::endl;
+  }
+
+private:
+  bool IsValid(const weak_t &in) const
+  {
+    return inuse_.count(in.key_);
+  }
+
+  void ReferenceCountedFree(const weak_t &in)
+  {
+    if (!IsValid(in))
+      return;
+    auto &pair = inuse_[in.key_];
+    --pair.second;
+    if (pair.second <= 0)
+    {
+      available_.push(pair.first);
+      inuse_.erase(in.key_);
+    }
+  }
+
+  void Free(const weak_t &in)
+  {
+    if (!IsValid(in))
+      return;
+    available_.push(inuse_[in.key_].first);
+    inuse_.erase(in.key_);
+  }
+
+  void AddCount(const weak_t &in)
+  {
+    if (!IsValid(in))
+      throw 1;
+    ++inuse_[in.key_].second;
+  }
+};
+
+// Lightly wraps a view of type T and holds a key that
+// that can be used to query if its storage
+// is in use or if it has been freed and allows
+// freeing of the storage.
+template <class T, class KEY_T>
+struct Pool<T, KEY_T>::weak_t : public T
```

Comment:
The `weak_t` is a non-owning type (naming inspired by `weak_ptr`) while `owner_t` is reference counted. 

My thinking on inheriting from `View` (or whatever object you are keeping in the pool) is that a pool member is an object of type `T`, so you should be able to treat it like one rather than accessing a member field.

Reply:
ah ok, yeah the naming makes sense. (I still don't love the inheritance, though it's not a big deal.)